# Ortholog tree

### Code

raxml_tree.sh

### in data

./tree_data

### output 

./output

![alt text](https://github.com/AsexGenomeEvol/Timema_asex_genomes/blob/main/3_Orthologs/trees/output/RAxML_bipartitions.five_pair_HOGs_genomeonly_1000boot.png)
Timema phylogeny. Maximum likelihood tree based on 2377398
orthologous coding DNA positions (from 3975 orthologs), rooted at the
midpoint. Branch lengths represent the mean number of substitutions per site. Node
labels indicate branch support (%) from 1000 bootstrap replicates. Othologs were
aligned using MCoffee (v11.00.8cbe486) (74) which was run with the following
aligners: mafft_msa, muscle_msa, clustalo_msa (75), and t_coffee_msa (76).
Alignments were concatenated together, and filtered with GBlocks (v. 0.91b, type =
codons, minimum block length = 12) to remove large alignment gaps and blocks of Ns
(113). The tree was then generated with RAxML (61), with a GTR+gamma model with
40 rate categories for each codon position.

