# Ortholog tree

### Code

raxml_tree.sh

### in data

./tree_data

### output 

./output

<img src="https://github.com/AsexGenomeEvol/Timema_asex_genomes/blob/main/3_Orthologs/trees/output/RAxML_bipartitions.five_pair_HOGs_all_1000boot_2.png"  width="500" />

**Timema phylogeny.** Maximum likelihood tree based on 2377398 orthologous coding DNA positions (from 3975 orthologs), rooted at the midpoint. Branch lengths represent the mean number of substitutions per site. Node labels indicate branch support (%) from 1000 bootstrap replicates. Othologs were aligned using MCoffee (v11.00.8cbe486) which was run with the following aligners: mafft_msa, muscle_msa, clustalo_msa, and t_coffee_msa. Alignments were concatenated together, and filtered with GBlocks (v. 0.91b, type = codons, minimum block length = 12) to remove large alignment gaps and blocks of Ns. The tree was then generated with RAxML, with a GTR+gamma model with 40 rate categories for each codon position.



